x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java

```diff
@@ -133,14 +140,14 @@ public CacheHelper getReaderCacheHelper() {
     }
 
     @SuppressForbidden(reason = "we manage references explicitly here")
-    private synchronized void onReaderClosed(IndexReader.CacheKey key) {
+    private synchronized void onReaderClosed(ElasticsearchDirectoryReader reader) {
```

Comment:
I'm unclear as to why we now need to have the reader itself rather than the cache key here?

Reply:
This is from an older iteration of the fix (where I found it easier to rewrite the code in terms of the reader than the cache key). I've reverted this part in f891f88ff1d